interface/src/app/signature/enriched_signatures.js

```diff
@@ -0,0 +1,216 @@
+    // care of by "<ml-model-validator>" component.
+    if (!$stateParams.mlmodel) {
+      return;
+    }
+
+    self.modelInUrl = $stateParams.mlmodel;
+    self.statusMessage = 'Connecting to the server ...';
+    self.pValueCutoff = 0.05;
+    self.enrichedSignatures = [];
+
+    // Do nothing if no genes are specified in URL.
+    if (!$stateParams.genes || !$stateParams.genes.split(',').length) {
+      self.statusMessage = 'No genes are specified.';
+      self.enrichedSignatures = [];
+      return;
+    }
+
+    var genesInUrl = [];
+    $stateParams.genes.split(',').forEach(function(token) {
+      var id = parseInt(token);
+      if (!isNaN(id) && genesInUrl.indexOf(id) === -1) {
+        genesInUrl.push(id);
+      }
+    });
+
+    // Promise that gets all signatures of given mlmodel in URL:
+    var apiPromises = [];
+    apiPromises.push(Signature.get(
+      {mlmodel: self.modelInUrl, limit: 0},
+      function success(response) {
+        self.signatures = Object.create(null);
+        response.objects.forEach(function(element) {
+          self.signatures[element.id] = element.name;
+        });
+      },
+      function error(err) {
+        var message = errGen('Failed to get signatures: ', err);
+        $log.error(message);
+        self.statusMessage = message + '. Please try again later.';
+      }
+    ).$promise);
+
+    var participations;
+    // Promise that gets all participation records that are related to
+    // the genes in URL:
+    apiPromises.push(Participation.get(
+      {'related_genes': $stateParams.genes, 'limit': 0},
+      function success(response) {
+        participations = response.objects;
+      },
+      function error(err) {
+        var message = errGen(
+          'Failed to get signature-gene participations: ', err);
+        $log.error(message);
+        self.statusMessage = message + '. Please try again later.';
+      }
+    ).$promise);
+
+    // Helper function that calculates the enrichment for each signature
+    // that includes at least one gene in the url.
+    var calculateEnrichments = function(typeName, genesBySignatures) {
+      var m = genesInUrl.length,
+        N = self.geneNum,
+        pValueArray = [],
+        matchedGenesBySignature = [],
+        signatureIDs = Object.keys(genesBySignatures);
+      signatureIDs.forEach(function(sigID) {
+        var genes = genesBySignatures[sigID],
+          n = Object.keys(genes).length, k = 0,
+          matchedGenes = [], selectedGene, pValue;
+        for (var i = 0; i < genesInUrl.length; i++) {
+          selectedGene = genesInUrl[i];
+          if (genes[selectedGene]) {
+            k++;
+            matchedGenes.push(genes[selectedGene]);
+          }
+        }
+        matchedGenesBySignature.push(matchedGenes);
+
+        // Four parameters are needed in the formula:
+        //     hyperGeometricTest(k, m, n, N)
+        // k: the number of heavy genes related to sigID in genesInUrl;
```

Comment:
@dongbohu - I think this `k` parameter is correct, in that it is the number of genes that are both in the `genesInUrl` and have some kind of participation in the desired signature. However, are we only doing "heavy" genes or all types of participation? Should this be reflected in the comments? Also, the `n` parameter below also mentions "heavy genes".